rspace/src/main/scala/coop/rchain/rspace/history/HistoryReader.scala

```diff
@@ -0,0 +1,81 @@
+package coop.rchain.rspace.history
+
+import cats.effect.Sync
+import coop.rchain.rspace.Blake2b256Hash
+import coop.rchain.rspace.internal._
+import coop.rchain.shared.Serialize
+import scodec.Codec
+
+/**
+  * Reader for particular history (state verified on blockchain)
+  *
+  * @tparam F effect type
+  * @tparam C type for Channel => this is Par
+  * @tparam P type for Pattern => this is BindPattern
+  * @tparam A type for Abstraction => this is ListParWithRandom
+  * @tparam K type for Continuation => this is TaggedContinuation
+  */
+trait HistoryReader[F[_], C, P, A, K, DK, JK, CK] {
+
+  // get content at a chanel
+  def getData(key: DK): F[Seq[Datum[A]]]
+
+  def getJoins(key: JK): F[Seq[Seq[C]]]
+
+  def getContinuations(key: CK): F[Seq[WaitingContinuation[P, K]]]
+
+  // get flavor for rich content (with raw ByteVector representation for each item)
+  def getRichDatums(key: DK): F[Seq[RichDatum[A]]]
+
+  def getRichJoins(key: JK): F[Seq[RichJoin[C]]]
+
+  def getRichContinuations(key: CK): F[Seq[RichKont[P, K]]]
+
+  // get current root which reader reads from
+  def root: Blake2b256Hash
+
+}
+
+/**
+  * Reader for history with addressing by Blake256Hash key
+  */
+trait HashHistoryReader[F[_], C, P, A, K]
+    extends HistoryReader[F, C, P, A, K, Blake2b256Hash, Blake2b256Hash, Blake2b256Hash] {
+
+  /**
+    * transform reader into [[RhoHistoryReader]]
+    *
+    * @param codecC codec for serializing channel
+    * @return
+    */
+  def toRho(
```

Comment:
Sorry for OOP fever, but thinking how to make it more pretty turned into a rabbit hole for me. Lets maybe arrange something later.

Reply:
Why is this called `rho`?